File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if s is None or len(s) == 0:
            return False
        
        st = []

        for i in range(len(s)):
            char = s[i]

            if char == "{":
                st.append("}")
            elif char == "[":
                st.append("]")
            elif char == "(":
                st.append(")")
            elif len(st) == 0 or st.pop() != char:
                return False
        
        
        return len(st) == 0
            

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if(s.length() == 0 || s == null){
            return false;
        }

        Stack<Character> st = new Stack<>();

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);

            if(ch == '{'){
                st.add('}');
            }
            else if(ch == '('){
                st.add(')');
            }
            else if(ch == '['){
                st.add(']');
            }else if(st.isEmpty() || st.pop() != ch){
                return false;
            }
        }

        return st.size() == 0;
    }
}

// Stack
// Time Complexity: O(n)
// Space Complexity: O(n)